### ISSUE TITLE
'Open In' non-UI Action extension not appearing in Share UI

### ISSUE THREAD

**Issue by realenginerd** [msintuneappsdk/ms-intune-app-sdk-ios#63]
I am trying to get my app to appear in the Share UI for certain (non-custom) file types. The intention is to allow our managed app to handle the file types we support within the Intune walled garden. We have an App Protection Policy such that apps in this policy can only share data with other apps in the same policy. My expectation is that when a user opens the Share menu in an app that is in this policy for the file types we support, our app should show up as an "Open In... " option, but unfortunately this does not happen.

I've followed the instructions in [Share Data via UIActivityViewController](https://docs.microsoft.com/en-us/intune/developer/app-sdk-ios#share-data-via-uiactivityviewcontroller). 

However, one thing I did not do is this:
> Your application must have at least one schemeURL defined under its Info.plist CFBundleURLTypes.

I'm confused as to why this is used here, as this is not required when setting up data sharing in a non-Intune world. It is also unclear as to what I should set my `CFBundleURLSchemes` to – our app does not support custom URL schemes (this is not required to set up data sharing in iOS), and the instructions do not indicate what URL scheme I should use. For these reasons, I was unable to complete that step in the setup. 

I've followed the rest of the instructions to the best of my ability, so I'm not sure what I am missing. Can someone please point me in the right direction? I am happy to provide more further information if needed. Thanks!

**Comment by aanavath** [comment]
Hi @realenginerd, a custom URL scheme allows users to open your app from other apps. The scheme needs to be unique, so you could use something with the name of your application. e.g. launch-MyApp (replacing MyApp with yours). 

This step is necessary for our method of implementing an alternate method of sharing where filter out apps unless they have Share/Action extensions.

**Comment by realenginerd** [comment]
Hello @aanavath. Following your advice, I've added the following to my app's Info.plist file:

```
<key>CFBundleURLTypes</key>
<array>
    <dict>
        <key>CFBundleTypeRole</key>
        <string>Viewer</string>
        <key>CFBundleURLName</key>
        <string>MyApp-intunemam</string>
        <key>CFBundleURLSchemes</key>
        <array>
            <string>MyApp-intunemam</string>
        </array>
    </dict>
</array>
```

(replacing my app's name with MyApp for privacy purposes). 

However, I still do not see my app appear in the list of apps to Open In to. It's only when we change the "Send org data to other apps" setting in the App Protection Policy from "Policy managed apps with Open-In/Share filtering" **to** "Policy managed apps" that our app appears in the Open In list.

**Comment by darosal** [comment]
Hi @realenginerd. Can you confirm that you have added the Intune types to your NSExtensionActivationRule according to the steps listed here: https://docs.microsoft.com/en-us/intune/developer/app-sdk-ios#update-existing-share-and-action-extensions? By querying for the Intune types, we filter out other apps that don't support them and, by extension, don't support Intune. If your app is appearing with "Policy managed apps" and not "Policy managed apps with Open-In/Share filtering", it is likely not accepting the Intune types in the NSExtensionActivationRule.

**Comment by realenginerd** [comment]
Hi @darosal, thanks for your reply! I noticed you linked to "Update existing Share and Action extensions". Unfortunately, our app does not have an existing Share or Action extension, so I've followed the rules listed in ['Copy To' Actions](https://docs.microsoft.com/en-us/intune/developer/app-sdk-ios#copy-to-actions) instead. Should we be following the instructions for existing Share and Action extensions **in addition** to the instructions listed in "`Copy To' Actions"?

In any case, our `NSExtensionActivationRule` is currently set to:
```
SUBQUERY (
    extensionItems,
    $extensionItem,
    SUBQUERY (
        $extensionItem.attachments,
        $attachment,
        ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "com.microsoft.intune.mam.public.data"
        || ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "com.microsoft.intune.mam.public.zip-archive"
    ).@count >= 1
).@count >= 1
```

The `NSExtensionActivationRule` looks right to me, so I'm not sure why our non-UI action extension doesn't appear in the share menu. Any other thoughts?

**Comment by darosal** [comment]
Hi @realenginerd, following the `NSExtensionActivationRule` step under the "Copy To" instructions should be sufficient. You mentioned that your app appears in the Share UI when the _Send Org data to other apps_ policy is set to "Policy managed apps". When this happens, does your app appear in the (1) top row of share apps or in the (2) column list of actions?
![Screenshot 2020-02-11 at 6 05 24 PM](https://user-images.githubusercontent.com/58273741/74280091-1107bf00-4cea-11ea-9140-e21251a348b5.png)

**Comment by realenginerd** [comment]
@darosal, our app appears on the top row (box 1 in the picture). 

I should clarify that in this scenario, it's the full app that appears, **not** the action extension. When I click on our app, it receives the file from the originating app with no issues. 

It doesn't matter to us whether our full app or the action extension appears on the list, as long as it is possible for our app to receive files from any Intune app with the *Send org data to other apps* policy set to 'Policy managed apps with Open-In/Share filtering'. Let me know if you have further questions!

**Comment by darosal** [comment]
@realenginerd, thanks for confirming that.

It might be helpful to explain a bit of the purpose behind the 'Copy To/Open In' extension. When an app includes values in 'CFBundleDocumentTypes' within the Info.plist, iOS will automatically create the Copy To "extension" and show it within the Share UI. Opening it through there will basically trigger an openURL with the scheme of your app.

When the _Send org data to other apps policy_ is set to  'Policy managed apps with Open-In/Share filtering', the Intune SDK filters out anything that doesn't support the Intune types, including these iOS generated Copy To  extensions. To bypass that and keep your app as a Copy To option, an app can add an action extension stub. Because this stub only accepts Intune types in the activation rule, it only shows when Open-In/Share filtering is on. When selected, the stub tells the requesting app to call an openURL with the URL scheme of your app.

Based on my understanding of the scenario you're facing, it seems to me that the extension itself is failing to appear. The Copy To extension that appears when 'Policy managed apps' is set is the one generated by iOS as a result of having values in 'CFBundleDocumentTypes'. However, the action extension stub never appears.

Can you try changing the NSExtensionActivationRule to the String 'TRUEPREDICATE' for debugging? If the action extension appears after that, then we'll know more of what's causing the issue. If that doesn't cause the extension to appear, there might be some misconfiguration of the extension itself that is breaking it. Make sure that both the extension and the main app have the same target iOS version.

**Comment by realenginerd** [comment]
@darosal, thanks for the detailed explanation! Always nice to understand what is going on behind the scenes.

It turns out the target iOS version for our extension was much newer than the target iOS version for our main app. Since our test device had an older version of iOS, the extension did not appear. Once we changed the target iOS version for the extension to match that of the main app, the extension started showing up in the share menu. We have progress!

However, the extension is not behaving as expected upon clicking it. If I am understanding correctly, I should expect a call to `openURL:` on the main app with a URL to the file, as if the "Copy to" button was present and had been clicked. However, the share menu disappears instead, and our main app is not called. Debugging the extension indicates that the `beginRequest(with context: NSExtensionContext)` function in ActionRequestHandler.swift is indeed called. Apart from making changes to the Action.js or ActionRequestHandler.swift files within the extension, which the instructions do not mention – what am I missing for the Intune SDK to implement the file sharing behavior?

**Comment by darosal** [comment]
@realenginerd, awesome! 

At this point, the copy to action not happening could be a result of a misconfiguration. But first, can you confirm that the action extension stub is being selected from a managed app? If the document is shared from the iOS Files app, nothing will happen. However if shared from a managed app like OneDrive, then it should complete the copy to if configured correctly.

If not, can you check the following?

- [ ] The SDK is added to the extension's 'Frameworks and Libraries'
- [ ] Both a 'MyApp' **and** 'MyApp-intunemam' scheme are listed under `CFBundleURLSchemes` in the main app's Info.plist
- [ ] **Both** the main app and the extension share an App Group under their capabilities **and both** have it listed in their Info.plist's `IntuneMAMSettings` in an `AppGroupIdentifiers` array 
- [ ] 'OpenInActionExtension' is set to 'YES' in the extension's Info.plist

**Comment by realenginerd** [comment]
Hi @darosal! From the items you've listed,

- **can you confirm that the action extension stub is being selected from a managed app**: yup, I've only used Outlook and OneDrive to share so far. 
- **The SDK is added to the extension's 'Frameworks and Libraries'**: yup
- **Both the main app and the extension share an App Group under their capabilities and both have it listed in their Info.plist's IntuneMAMSettings in an AppGroupIdentifiers array**: yup
- **'OpenInActionExtension' is set to 'YES' in the extension's Info.plist**: yup
- **Both a 'MyApp' and 'MyApp-intunemam' scheme are listed under CFBundleURLSchemes in the main app's Info.plist**: ahh, so *this* is what I am missing. The instructions say "Your application must have at least one schemeURL defined under its Info.plist `CFBundleURLTypes`.", but it does not specify *which* types to add. This makes it explicit what scheme to add.

However, I am still unable to share data to my app after adding the `MyApp` and `MyApp-intunemam` schemes to the main app. I've added the schemes in my Info.plist like so:

```
<key>CFBundleURLTypes</key>
    <array>
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Viewer</string>
            <key>CFBundleURLName</key>
            <string>MyApp</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>MyApp</string>
            </array>
        </dict>
        <dict>
            <key>CFBundleTypeRole</key>
            <string>Viewer</string>
            <key>CFBundleURLName</key>
            <string>MyApp-intunemam</string>
            <key>CFBundleURLSchemes</key>
            <array>
                <string>MyApp-intunemam</string>
            </array>
        </dict>
    </array>
```

but when I click on the 'Open In MyApp' extension, I get an "Action Not Allowed: Your organization only allows you to transfer its data to other managed apps" error message. In addition, the swift code on the extension is no longer called. 

What else am I missing? Was there some other part of Intune I needed to add these schemes to? Some additional configuration steps?

**Comment by darosal** [comment]
@realenginerd, I'll have to update the docs to clarify that step. 

The error message you saw pops up when there is no managed account signed into the extension's host app. Can you check that? If it still pops up even after that, we'll have to do some more troubleshooting.

**Comment by realenginerd** [comment]
@darosal, by 'signed in' did you mean call `loginAndEnrollAccount()` on IntuneMAMEnrollmentManager? If so then yeah, I've already signed in. Here's my workflow:
1. Launch the debugger for the extension (At this point, Outlook opens.)
1. Switch to the main app and log in
1. Switch back to Outlook to share the file
1. Click on "Open in Extension"

Our main app performs the login to Intune after its login to our own services. (Our app is not integrated with ADAL at this time, but we still link to the ADAL framework.) At step 2, I navigate back into the app to log in, to ensure the user is logged in to Intune. The call to `loginAndEnrollAccount()` here returns a 208 because this user has already been enrolled.

**Comment by darosal** [comment]
@realenginerd, I see.

If the account is properly targeted and receiving policies, then the action extension is likely unable to access that information. Do both the action extension and the main app have 'com.microsoft.intune.mam' listed under the Keychain Sharing capability? 

If they do, are both the main app and the action extension signed with the same keychain-access-groups? Do the following to find out: 
- Find the provisioning profile used to sign each in Finder. 
- Select them each and check 'Get Info'
- Look under 'Preview:'
- Scroll to the 'ENTITLEMENTS' section
- Check the 'keychain-access-groups' setting

**Comment by realenginerd** [comment]
Oh my, @darosal, I think we have a breakthrough!

I added Keychain Sharing to the action extension, and added 'com.microsoft.intune.mam' to the list of keychain groups. I also confirmed that the main app's provisioning profile *and* the action extension's provisioning profile are both signed with the same keychain-access-group. After that, I was able to send files from Intune apps (Outlook and OneDrive in my testing) to my app!

I want to note, however, that two items were lacking in the [documentation](https://docs.microsoft.com/en-us/intune/developer/app-sdk-ios#copy-to-actions) and prevented me from being able to implement this file sharing functionality immediately: 1) this step of adding the keychain group to the action extension and 2) instruction regarding which `CFBundleURLScheme` to add. I would highly recommend that they be added to the documentation.

Thanks so much for your time and help!
